ft: ZENKO-1175 tailable cursor to consume mongo oplog

jonathan-gramain opened the pull request:
Use a tailable custor to keep ordering guarantees for the records we
read. This also means we have to read from the beginning when we
reconnect (at startup), and start processing when we encountered the
unique ID previously stored in zookeeper.

Also removed dispatcher mode with MongoLogReader (was only used for
the short-lived Federation deployment of Zenko).

jonathan-gramain committed:
ft: ZENKO-1175 tailable cursor to consume mongo oplog

Use a tailable custor to keep ordering guarantees for the records we
read. This also means we have to read from the beginning when we
reconnect (at startup), and start processing when we encountered the
unique ID previously stored in zookeeper.

Also removed dispatcher mode with MongoLogReader (was only used for
the short-lived Federation deployment of Zenko).

bert-e commented:
# Hello jonathan-gramain,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get
information on this process.





*Status report is not available.*

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `bugfix/ZENKO-1175-oplogSkipFix` into
`development/8.0`
* [w/8.1/bugfix/ZENKO-1175-oplogSkipFix](https://github.com/scality/arsenal/compare/development/8.1...w/8.1/bugfix/ZENKO-1175-oplogSkipFix) will be merged into `development/8.1` (pull request #595)



The following branches will **NOT** be impacted:

* `development/6.4`
* `development/7.4`




*Follow* integration pull requests if you would like to be notified of
build statuses by email.

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:



* 2 peers





Peer approvals *must* include at least 1 approval from the following list:

* @RachedBenMustapha

* @rahulreddy

* @bennettbuchanan

philipyoo commented on lib/storage/metadata/mongoclient/ListRecordStream.js:
```diff
@@ -1,32 +1,42 @@
         // this._unpublishedListing is true once we pass the oplog
-        // that has the start seq timestamp and uniqID 'h'
+        // record that has the same uniqID 'h' than last saved. If we
+        // don't find it (e.g. log rolled over before populator could
```
just curious, when does the log roll? Does this mean some entries may get lost?

jonathan-gramain replied:
The oplog is a "capped collection", MongoDB uses it internally as such, here's the MongoDB doc description:

> Capped collections are fixed-size collections that support high-throughput operations that insert and retrieve documents based on insertion order. Capped collections work in a way similar to circular buffers: once a collection fills its allocated space, it makes room for new documents by overwriting the oldest documents in the collection.

So the log rolls when it runs out of space (so space-based rather than time-based), and it indeed means oldest entries will get overwritten by newer ones.

The default sizing seems to be 5% of available storage space, with a minimum of 990MB (https://docs.mongodb.com/v3.4/core/replica-set-oplog/#replica-set-oplog-sizing)

For this reason at least, we will keep the "sweeper" in case old entries are overwritten before they are being processed, either as a manually-triggered task or as an automatically scheduled periodic task.

philipyoo replied:
nice, thank you for the info!

jonathan-gramain commented:
@bert-e wait

ghost commented:
There were the following issues with this Pull Request

* Commit: 4e20c1c78cb3d401a6ce37531ba7dab820c91ac7
  - ✖ message may not be empty
  - ✖ type may not be empty
* Commit: 77ec22755237f794227b3d8377120833628e9037
  - ✖ message may not be empty
  - ✖ type may not be empty
* Commit: 6d57912c50e6aaf83d45db6272d397031335559b
  - ✖ message may not be empty
  - ✖ type may not be empty


You may need to [change the commit messages][ref] to comply with the repository contributing guidelines.

--------

🤖 This comment was generated by [**commitlint[bot]**][repo]. Please report issues [here][issues].

Happy coding!

[ref]: https://help.github.com/articles/changing-a-commit-message/
[repo]: https://github.com/z0al/commitlint-bot
[issues]: https://github.com/z0al/commitlint-bot/issues

ghost commented:
There were the following issues with this Pull Request

* Commit: 4e20c1c78cb3d401a6ce37531ba7dab820c91ac7
  - ✖ message may not be empty
  - ✖ type may not be empty
* Commit: 77ec22755237f794227b3d8377120833628e9037
  - ✖ message may not be empty
  - ✖ type may not be empty
* Commit: 6d57912c50e6aaf83d45db6272d397031335559b
  - ✖ message may not be empty
  - ✖ type may not be empty
* Commit: 83add92383ce2ef89b884847df70b15d7f80b2bc
  - ✖ message may not be empty
  - ✖ type may not be empty


You may need to [change the commit messages][ref] to comply with the repository contributing guidelines.

--------

🤖 This comment was generated by [**commitlint[bot]**][repo]. Please report issues [here][issues].

Happy coding!

[ref]: https://help.github.com/articles/changing-a-commit-message/
[repo]: https://github.com/z0al/commitlint-bot
[issues]: https://github.com/z0al/commitlint-bot/issues

ghost commented:
There were the following issues with this Pull Request

* Commit: 89647e076f97ed9bdd9dfbc35365237ffaef6d2f
  - ✖ message may not be empty
  - ✖ type may not be empty


You may need to [change the commit messages][ref] to comply with the repository contributing guidelines.

--------

🤖 This comment was generated by [**commitlint[bot]**][repo]. Please report issues [here][issues].

Happy coding!

[ref]: https://help.github.com/articles/changing-a-commit-message/
[repo]: https://github.com/z0al/commitlint-bot
[issues]: https://github.com/z0al/commitlint-bot/issues

jonathan-gramain commented:
Note for reviewers: since last status, I have removed the use of cursor stream() in favor of calling the next() function directly, because the cursor stream interface was not always able to report errors happening on the mongodb connection correctly.

vrancurel commented on lib/storage/metadata/mongoclient/LogConsumer.js:
```diff
@@ -18,12 +17,9 @@ class LogConsumer {
-        this.oplogNsRegExp = new RegExp(`^${database}\\.`);
+        this._mongoUrl = `mongodb://${replicaSetHosts}/local`;
+        this._logger = logger;
+        this._oplogNsRegExp = new RegExp(`^${database}\\.`);
```
what does it bring to underscore mongoUrl, logger and oplogNsRegExp ? because it makes a bigger changeset, which is therefore harder to read.

ploki replied:
Maybe it's a convention to inform that these members must not be used from outside of the class

jonathan-gramain replied:
They are private variable, to mark them as such (per coding guidelines we use) they should start with an underscore

vrancurel commented on lib/storage/metadata/mongoclient/ListRecordStream.js:
```diff
@@ -1,33 +1,62 @@
+        // process its oldest entries), we will restart from the
+        // latest record of the oplog.
         this._unpublishedListing = false;
+        this._skipCount = 0;
```
Please comment what skipCount does

ghost commented:
There were the following issues with this Pull Request

* Commit: 89647e076f97ed9bdd9dfbc35365237ffaef6d2f
  - ✖ message may not be empty
  - ✖ type may not be empty
* Commit: 858f583e6f02d9a36cec982697f736a885f93733
  - ✖ message may not be empty
  - ✖ type may not be empty


You may need to [change the commit messages][ref] to comply with the repository contributing guidelines.

--------

🤖 This comment was generated by [**commitlint[bot]**][repo]. Please report issues [here][issues].

Happy coding!

[ref]: https://help.github.com/articles/changing-a-commit-message/
[repo]: https://github.com/z0al/commitlint-bot
[issues]: https://github.com/z0al/commitlint-bot/issues

ghost commented:
There were the following issues with this Pull Request

* Commit: 89647e076f97ed9bdd9dfbc35365237ffaef6d2f
  - ✖ message may not be empty
  - ✖ type may not be empty
* Commit: 858f583e6f02d9a36cec982697f736a885f93733
  - ✖ message may not be empty
  - ✖ type may not be empty
* Commit: f2f1bdd9769c774d4211ff54bcd37d7c440b431a
  - ✖ message may not be empty
  - ✖ type may not be empty


You may need to [change the commit messages][ref] to comply with the repository contributing guidelines.

--------

🤖 This comment was generated by [**commitlint[bot]**][repo]. Please report issues [here][issues].

Happy coding!

[ref]: https://help.github.com/articles/changing-a-commit-message/
[repo]: https://github.com/z0al/commitlint-bot
[issues]: https://github.com/z0al/commitlint-bot/issues

ghost commented:
There were the following issues with this Pull Request

* Commit: 89647e076f97ed9bdd9dfbc35365237ffaef6d2f
  - ✖ message may not be empty
  - ✖ type may not be empty
* Commit: 858f583e6f02d9a36cec982697f736a885f93733
  - ✖ message may not be empty
  - ✖ type may not be empty
* Commit: f2f1bdd9769c774d4211ff54bcd37d7c440b431a
  - ✖ message may not be empty
  - ✖ type may not be empty
* Commit: 7056c65ddaaf75f8ecccdee0607f7fe961701b59
  - ✖ message may not be empty
  - ✖ type may not be empty


You may need to [change the commit messages][ref] to comply with the repository contributing guidelines.

--------

🤖 This comment was generated by [**commitlint[bot]**][repo]. Please report issues [here][issues].

Happy coding!

[ref]: https://help.github.com/articles/changing-a-commit-message/
[repo]: https://github.com/z0al/commitlint-bot
[issues]: https://github.com/z0al/commitlint-bot/issues

bert-e commented:
# History mismatch



Merge commit #829eba47e995b5e0bfdb569580c05ea2f3184867 on the integration branch
`w/8.1/bugfix/ZENKO-1175-oplogSkipFix` is merging a branch which is neither the current
branch `bugfix/ZENKO-1175-oplogSkipFix` nor the development branch
`development/8.1`.

It is likely due to a rebase of the branch `bugfix/ZENKO-1175-oplogSkipFix` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

jonathan-gramain commented:
@bert-e reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `bugfix/ZENKO-1175-oplogSkipFix` into
`development/8.0`
* [w/8.1/bugfix/ZENKO-1175-oplogSkipFix](https://github.com/scality/arsenal/compare/development/8.1...w/8.1/bugfix/ZENKO-1175-oplogSkipFix) will be merged into `development/8.1` (pull request #599)



The following branches will **NOT** be impacted:

* `development/6.4`
* `development/7.4`




*Follow* integration pull requests if you would like to be notified of
build statuses by email.

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:



* 2 peers





Peer approvals *must* include at least 1 approval from the following list:

* @RachedBenMustapha

* @rahulreddy

* @bennettbuchanan

jonathan-gramain commented:
@bert-e wait

jonathan-gramain committed:
bf: ZENKO-1175 fix when no saved ID exists

In case where there is no saved ID yet (initial deployment), do
process the very first entry in the log instead of skipping it. In
practice it should not have an impact because the very first entry in
the log is normally not due to be processed for CRR, but it ensures
correctness.

ghost commented:
There were the following issues with this Pull Request

* Commit: 572b94dc626bc706ef439d6f529e18490fb8b850
  - ✖ message may not be empty
  - ✖ type may not be empty
* Commit: b6bc9984622f3a2910fbadad95ae91ead027166f
  - ✖ message may not be empty
  - ✖ type may not be empty


You may need to [change the commit messages][ref] to comply with the repository contributing guidelines.

--------

🤖 This comment was generated by [**commitlint[bot]**][repo]. Please report issues [here][issues].

Happy coding!

[ref]: https://help.github.com/articles/changing-a-commit-message/
[repo]: https://github.com/z0al/commitlint-bot
[issues]: https://github.com/z0al/commitlint-bot/issues

rahulreddy reviewed (approved):
My only concern was the skipping of first entry. Now that it's fixed, LGTM.

rahulreddy commented:
One more thing, please update `package.json`'s version attribute by incrementing the patch version.

bert-e commented:
# History mismatch



Merge commit #a849606688e615df334d03282151b5761d19ec03 on the integration branch
`w/8.1/bugfix/ZENKO-1175-oplogSkipFix` is merging a branch which is neither the current
branch `bugfix/ZENKO-1175-oplogSkipFix` nor the development branch
`development/8.1`.

It is likely due to a rebase of the branch `bugfix/ZENKO-1175-oplogSkipFix` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

jonathan-gramain commented:
@bert-e reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `bugfix/ZENKO-1175-oplogSkipFix` into
`development/8.0`
* [w/8.1/bugfix/ZENKO-1175-oplogSkipFix](https://github.com/scality/arsenal/compare/development/8.1...w/8.1/bugfix/ZENKO-1175-oplogSkipFix) will be merged into `development/8.1` (pull request #600)



The following branches will **NOT** be impacted:

* `development/6.4`
* `development/7.4`




*Follow* integration pull requests if you would like to be notified of
build statuses by email.

bert-e commented:
# Build failed



The build for [commit](https://github.com/scality/Arsenal/commit/7f5413699dfdbc29cb08806a41312e9a12ff11eb) did not succeed in branch w/8.1/bugfix/ZENKO-1175-oplogSkipFix.

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/8.0`

* :heavy_check_mark: `development/8.1`



The following branches will **NOT** be impacted:

* `development/6.4`
* `development/7.4`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/8.0`

* :heavy_check_mark: `development/8.1`



The following branches have **NOT** changed:

* `development/6.4`
* `development/7.4`



Please check the status of the associated issue ZENKO-1175.

Goodbye jonathan-gramain.
